dxr/plugins/clang/dxr-index.cpp

```diff
@@ -1245,34 +1280,41 @@ class DXRIndexAction : public PluginASTAction {
       D.Report(DiagID) << output;
       return false;
     }
-    output = realpath(output.c_str(), NULL);
+    output = abs_output;
     output += "/";
+    FileInfo::output = output;
 
-    const char* tmp = getenv("DXR_CXX_CLANG_TEMP_FOLDER");
-    if(tmp){
+    // The temp directory for this plugin's output.
+    const char *tmp = getenv("DXR_CXX_CLANG_TEMP_FOLDER");
+    std::string tmpdir;
+    if (tmp)
       tmpdir = tmp;
-    }else
+    else
       tmpdir = output;
-
-    char* abs_tmpdir = realpath(tmpdir.c_str(), NULL);
-    if(!abs_tmpdir){
+    char *abs_tmpdir = realpath(tmpdir.c_str(), nullptr);
+    if (!abs_tmpdir) {
       DiagnosticsEngine &D = CI.getDiagnostics();
       unsigned DiagID = D.getCustomDiagID(DiagnosticsEngine::Error,
         "Temporary directory '%0' does not exist");
       D.Report(DiagID) << tmpdir;
       return false;
     }
-    tmpdir = realpath(tmpdir.c_str(), NULL);
+    tmpdir = abs_tmpdir;
     tmpdir += "/";
+    IndexConsumer::setTmpDir(tmpdir);
+
+    free(abs_src);
```

Comment:
> Actually, I don't think we should free any of those long-lived vars in ParseArgs at all (even when they go out of scope): the memory they point to must remain unmolested the whole time the AST walker runs, since we're just copying the ptrs into the static vars, and ParseArgs is long exited by the time the AST walker even starts.

Thanks for looking all this over!  I probably won't get to your other feedback until tomorrow or Friday, but I think on this one that the assignment actually copies out the underlying char data, not just the pointer:
http://en.cppreference.com/w/cpp/string/basic_string/operator%3D
I did a little stub test with the same setup as in dxr-index.cpp that seems to confirm that: after the `FileInfo::srcdir = abs_src;` assignment, `abs_src` and `FileInfo::srcdir.data()` point to different data.